File: ball_chaser/ball_chaser/drive_bot.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from robot_interfaces.srv import DriveToTarget

class DriveBot(Node):
    def __init__(self):
        super().__init__("drive_bot")
        self.get_logger().info("Hello drive bot")
        
        # create publisher
        self.motor_command_publisher = self.create_publisher(Twist, "/cmd_vel", 10)

        # * 建立 Service ( for service only )
        self.create_service(DriveToTarget, "ball_chaser/command_robot", self.handle_drive_request)

    # * 建立 call back function, 撰寫執行動作
    def handle_drive_request(self, request, response):
        self.get_logger().info("DriveToTargetRequest received - linear_x: " \
            + str(request.linear_x) + ", angular_z: " + str(request.angular_z))

        # setup transer message
        motor_command = Twist()
        motor_command.linear.x = request.linear_x
        motor_command.angular.z = request.angular_z
        # publish message to /cmd_vel topic
        self.motor_command_publisher.publish(motor_command)
        # setup return message
        response.msg_feedback = "OK"
        return response


def main(args=None):
    #  初始化 ROS
    rclpy.init(args=args)
    #  建立 Node
    node = DriveBot()
    #  Spin Node 以持續運行Node工作(執行 callback function)
    rclpy.spin(node)
    #  Shutdown ROS  
    rclpy.shutdown()

if __name__=='__main__':
    main()

File: ball_chaser_cpp/src/drive_bot.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "robot_interfaces/srv/drive_to_target.hpp"

using std::placeholders::_1;
using std::placeholders::_2;
using namespace std;
using namespace geometry_msgs::msg;
using namespace robot_interfaces::srv;

class DriveBot : public rclcpp::Node
{
public:
    DriveBot() : Node("drive_bot") 
    {
        RCLCPP_INFO(this->get_logger(), "init Drive Bot");
        service = this->create_service<DriveToTarget>(
                "/ball_chaser/command_robot", 
                std::bind(&DriveBot::handle_drive_request, this, _1, _2));
    }

private:
    rclcpp::Publisher<Twist>::SharedPtr motor_command_publisher;
    rclcpp::Service<DriveToTarget>::SharedPtr service;

    void handle_drive_request(const DriveToTarget::Request::SharedPtr req, 
            DriveToTarget::Response::SharedPtr res)
    {
        RCLCPP_INFO(this->get_logger(), "DriveToTargetRequest received - linear_x: %1.2f, angular_z: %1.2f", 
                req->linear_x, req->angular_z);

        Twist motor_command;
        // auto motor_command = std::make_shared<geometry_msgs::msg::Twist>();

        motor_command.linear.x = req->linear_x;
        motor_command.angular.z = req->angular_z;

        motor_command_publisher = this->create_publisher<Twist>("/cmd_vel", 10);
        motor_command_publisher->publish(motor_command);

        res->msg_feedback = "Motor command received -linear_x "+ to_string(motor_command.linear.x) 
                + ", angular_z: "+to_string(motor_command.angular.z);
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = make_shared<DriveBot>();
    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;
}

